# **2조**
**신경망 (활성화함수(시그모이드, ReLU등), 다차원 배열의 계산, 다층 신경망 구현, 출력층 설계)**

## 1.활성화 함수

1. 활성화 함수를 비선형 함수로 구현하는 이유를 간략히 설명해주세요. (4점)

설명: 단층 퍼셉트론이 XOR 문제와 같은 비선형 문제를 해결할 수 없는 한계를 극복하는 등의 장점이 있어 사용한다.

아래 코드를 실행해주세요.

In [1]:
import torch
import matplotlib.pyplot as plt

2. 아래 시그모이드 함수 구현 코드(??? 부분)를 완성하고 시그모이드 함수에 대해서 간략히 설명해주세요. (4점)

In [6]:
x = torch.arange(-5., 5., 0.1)
y = torch.sigmoid(x)
plt.plot(x.numpy(), y.numpy())
plt.show()

RuntimeError: Numpy is not available

설명: 시그모이드 함수는 입력값을 0과 1 사이의 값으로 변환하는 비선형 활성화 함수입니다. 

3. 아래 하이퍼볼릭탄젠트 함수 구현 코드를 완성하고 하이퍼볼릭탄젠트 함수에 대해서 간략히 설명해주세요. (4점)

In [7]:
x = torch.range(-5., 5., 0.1)
y = torch.tanh(x)
plt.plot(x.numpy(), y.numpy())
plt.show()

C:\Users\Public\Documents\ESTsoft\CreatorTemp\ipykernel_6148\612086185.py:1: UserWarning: torch.range is deprecated and will be removed in a future release because its behavior is inconsistent with Python's range builtin. Instead, use torch.arange, which produces values in [start, end).
  x = torch.range(-5., 5., 0.1)


RuntimeError: Numpy is not available

설명:하이퍼볼릭 탄젠트 함수(tanh 함수)는 또 다른 비선형 활성화 함수로, 시그모이드 함수와 유사하지만 출력 값이 -1과 1 사이에 있습니다. 

4. 시그모이드 함수를 사용할 때 나타나는 그래디언트 소실 문제에 대해 산략히 설명해주세요. (4점)

설명:그래디언트 소실 문제는 신경망의 역전파 과정에서 발생하는 문제로, 특히 깊은 신경망에서 자주 발생합니다. 

5. 아래 렐루 함수 구현 코드를 완성하고 렐루 함수의 장점과 단점에 대해서 간략히 설명해주세요. (4점)

In [8]:
relu = torch.nn.ReLU()
x = torch.range(-5., 5., 0.1)
y=relu(x)
plt.plot(x.numpy(), y.numpy())
plt.show()

C:\Users\Public\Documents\ESTsoft\CreatorTemp\ipykernel_6148\1463588837.py:2: UserWarning: torch.range is deprecated and will be removed in a future release because its behavior is inconsistent with Python's range builtin. Instead, use torch.arange, which produces values in [start, end).
  x = torch.range(-5., 5., 0.1)


RuntimeError: Numpy is not available

설명: 

6. 아래 소프트맥스 함수 구현 코드를 완성하고 소프트맥스 함수에 대해서 간략히 설명해주세요. (4점)

In [ ]:
softmax = torch.nn.softmax(dim=1)
x_input = torch.randn(1,3)
y_output = softmax(x_input)
print(x_input)
print(y_output)
print(torch.sum(y_output, dim=1))

설명: 입력받은 값들을 각 클래스에 대한 확률값으로 변환해줍니다.

## 2. 다차원배열계산, 다층신경망

1. 아래 2차원 텐서를 구현하는 코드를 완성해주세요. (3점)

In [9]:
tensor_torch1 = torch.tensor([[1, 2, 3], [4, 5, 6]])
tensor_torch2 = torch.tensor([[1, 2, 3], [4, 5, 6]])

2. 텐서 간의 곱셈, 행렬곱셈, (3,2)로의 차원변경을 수행하는 코드를 완성해주세요. (3점)

In [10]:
tensor_mul_torch = torch.mul(tensor_torch1, tensor_torch2)
tensor_mul_torch

tensor([[ 1,  4,  9],
        [16, 25, 36]])

In [16]:
tensor_matmul_torch = torch.matmul(tensor_torch1, tensor_torch2.view(3,2))
tensor_matmul_torch

tensor([[22, 28],
        [49, 64]])

3. 아래 다층신경망의 데이터 준비 코드를 실행하고 코드에 대해 자유롭게 설명해주세요. (5점)

In [ ]:
import torch
from torch import nn, optim
from torchvision import datasets, transforms
from torch.utils.data import DataLoader

# 데이터셋 변환 정의
transform = transforms.Compose([transforms.ToTensor(), transforms.Normalize((0.5,), (0.5,))])

# MNIST 데이터셋 로드
train_dataset = datasets.MNIST(root='./data', train=True, download=True, transform=transform)
test_dataset = datasets.MNIST(root='./data', train=False, download=True, transform=transform)

# DataLoader 정의
train_loader = DataLoader(train_dataset, batch_size=64, shuffle=True)
test_loader = DataLoader(test_dataset, batch_size=64, shuffle=False)

설명: MNIST 데이터셋 가져와서 데이터셋 준비하는 과정

4. 아래 다층신경망의 모델 정의 코드를 실행하고 코드에 대해 자유롭게 설명해주세요. (5점)

In [ ]:
class MLP(nn.Module):
    def __init__(self):
        super(MLP, self).__init__()
        self.fc1 = nn.Linear(28*28, 256)
        self.fc2 = nn.Linear(256, 128)
        self.fc3 = nn.Linear(128, 64)
        self.fc4 = nn.Linear(64, 10)

    def forward(self, x):
        x = x.view(-1, 28*28)
        x = torch.relu(self.fc1(x))
        x = torch.relu(self.fc2(x))
        x = torch.relu(self.fc3(x))
        x = self.fc4(x)
        return x

model = MLP()
criterion = nn.CrossEntropyLoss()
optimizer = optim.Adam(model.parameters(), lr=0.001)

설명: 다층 신경망 모델을 정의하고, 이를 통해 MNIST 데이터셋을 분류하는 준비를 마칩니다. 모델은 네 개의 fully connected 레이어로 구성되어 있으며, 각 레이어 사이에는 ReLU 활성화 함수가 적용됩니다

5. 아래 다층신경망의 모델 훈련 코드를 실행하고 코드에 대해 자유롭게 설명해주세요. (실행 시간이 오래 걸립니다) (5점)

In [ ]:
epochs = 5

for epoch in range(epochs):
    running_loss = 0
    for images, labels in train_loader:
        # 모델 초기화
        optimizer.zero_grad()

        # 순전파
        outputs = model(images)
        loss = criterion(outputs, labels)

        # 역전파
        loss.backward()
        optimizer.step()

        running_loss += loss.item()

    print(f"Epoch {epoch+1}, Loss: {running_loss/len(train_loader)}")

설명: 모델을 불러와 5epoch 학습하는 코드입니다.

6. 아래 다층신경망의 모델 평가 코드를 실행하고 코드에 대해 자유롭게 설명해주세요. (5점)

In [ ]:
correct = 0
total = 0

with torch.no_grad():
    for images, labels in test_loader:
        outputs = model(images)
        _, predicted = torch.max(outputs.data, 1)
        total += labels.size(0)
        correct += (predicted == labels).sum().item()

print(f"Accuracy: {100 * correct / total}%")

설명:학습된 모델의 성능을 평가하고, 모델이 테스트 데이터셋에서 얼마나 잘 동작하는지를 판단하는 데 유용합니다.

---

# **3조**
**신경망 학습(손실 함수, 수치 미분(편미분), 기울기(경사 하강법))**

##1. 손실함수와 수치미분(편미분) - 총 25점


### **1.1) 다음의 문장을 완성하시오.(5점, 각 1점씩)**

손실 함수는 신경망 성능의 '???'을 나타내는 지표로, 현재 신경망이 훈련 데이터를 얼마나 잘 처리하지 '???'하느냐를 나타냅니다.

손실 함수는 신경망 성능의 '???'을 나타내는 지표로, 현재 신경망이 훈련 데이터를 얼마나 잘 처리하지 '???'하느냐를 나타냅니다.

 가장 많이 쓰이는 손실 함수는 크로스엔트로피 손실 함수입니다.

평균 제곱 오차는 각 원소의 출력(추정 값)과 정답 레이블(참 값)의 차를 제곱한 후, 그 평균을 구합니다.

원소의 출력(추정 값)과 정답 레이블(참 값)의 차가 작으면 작을 수록 ???.

### **1.2)  *평균 제곱 오차* 함수를 완성하시오. (5점)**

In [17]:
import numpy as np

def mean_squared_error(y, t):
    return 0.5 * np.mean((y - t)**2)


t = [0, 0, 1, 0, 0, 0, 0, 0, 0, 0]
y1 = [0.1, 0.05, 0.6, 0.0, 0.05, 0.1, 0.0, 0.1, 0.0, 0.0]

print( mean_squared_error(np.array(y1), np.array(t)) )
# >>> 0.0975000000031


y2 = [0.1, 0.05, 0.1, 0.0, 0.05, 0.1, 0.0, 0.6, 0.0, 0.0]
print( mean_squared_error(np.array(y2), np.array(t)) )
# >>> 0.597500000003

0.009750000000000003
0.059750000000000004


모범 답안)<br>
0.09750000000000003<br>
0.5975


### **1.3)   *교차 엔트로피 오차* 함수를 완성하시오. (5점)**

In [18]:
def cross_entropy_error(y, t):
    delta = 1e-7
    return -np.sum(t * np.log(y + delta))



t = [0, 0, 1, 0, 0, 0, 0, 0, 0, 0]
y = [0.1, 0.05, 0.6, 0.0, 0.05, 0.1, 0.0, 0.1, 0.0, 0.0]

print( cross_entropy_error(np.array(y1), np.array(t)) )
 # >>> 0.5108254709933802


y = [0.1, 0.05, 0.1, 0.0, 0.05, 0.1, 0.0, 0.6, 0.0, 0.0]

print( cross_entropy_error(np.array(y2), np.array(t) ))
 # >>> 2.3025840929945458

0.510825457099338
2.302584092994546


모범 답안)<br>
0.510825457099338<br>
2.302584092994546

### **1.4) 신경망 학습에서 손실 함수를 설정하는 이유를 서술하시오. (5점)**

답안) 학습 과정에서 모델이 얼마나 잘 동작하는지를 평가하고, 이를 통해 모델을 최적화하기 위한 방향을 제시하기 위함입니다.

### **1.5) 다음의 문장을 완성하시오.(5점)**

**수치미분이란?**

 '해석적 미분'은 우리가 수학 시간에 배운 오차를 포함하지 않는 '진정한 미분' 값을 구해줍니다. '수치 미분'은 이를 '근사적'로 계산하는 방법입니다. 따라서 오차가 포함됩니다.


**편미분이란?**

 편미분은 변수가 여러 개인 함수에 대한 미분을 의미합니다.

##2. 기울기 (경사 하강법) - 총 25점

### **2.1) 경사 하강법(Gradient Descent)이란 무엇인가요? 간략하게 서술해주세요. (3점)**

답안) 경사 하강법은 현재 위치에서의 기울기(경사)를 계산하여 그 기울기가 가리키는 방향으로 조금씩 이동하면서 손실 함수의 최솟값을 찾는 방법입니다. 

### **2.2) 경사 하강법의 대표적인 3가지 종류에 대해 각각 설명하고, 차이점을 설명하세요. (5점)**

배치 경사 하강법: 배치 경사 하강법은 한 번의 파라미터 업데이트를 위해 전체 훈련 데이터셋을 사용합니다.

확률적 경사 하강법: 매 반복마다 하나의 훈련 데이터 포인트를 사용하여 그래디언트를 계산합니다.

미니배치 경사 하강법: 미니배치 경사 하강법은 배치 경사 하강법과 확률적 경사 하강법의 절충안입니다.

### **2.3) 단순 선형 회귀를 위한 Full batch 경사하강법 예제 코드입니다. 빈칸을 채워주세요. (5점)**

In [19]:
import numpy as np

X = 2 * np.random.rand(100, 1)
y = 4 + 3 * X + np.random.randn(100, 1)

theta = np.random.randn(2, 1)
learning_rate = 0.1
iterations = 1000
m = len(X)

def gradient_descent(X_b, y, theta, learning_rate, iterations):
    for iteration in range(iterations):
        gradients = 2/m * X_b.T.dot(X_b.dot(theta) - y)
        theta = theta - learning_rate * gradients
    return theta

X_b = np.c_[np.ones((m, 1)), X]

theta_final = gradient_descent(X_b, y, theta, learning_rate, iterations)

print("최종 theta 값:", theta_final)

최종 theta 값: [[3.88738081]
 [3.12221329]]


모범 답안)<br>
최종 theta 값: [[3.79452122]<br>
 [3.1510319 ]]


### **2.4) 단순 선형 회귀를 위한 Stochastic 경사하강법 예제 코드입니다. 빈칸을 채워주세요. (5점)**

In [20]:
import numpy as np

X = np.array([1, 2, 3, 4, 5])
y = np.array([2, 4, 6, 8, 10])

m, b = 0, 0
learning_rate = 0.01
iterations = 1000

for _ in range(iterations):
    for i in range(len(X)):
        y_pred = m * X[i] + b
        error = y_pred - y[i]
        m_grad = 2 * X[i] * error
        b_grad = 2 * error
        m -= learning_rate * m_grad
        b -= learning_rate * b_grad

print(f"m = {m}, b = {b}")

m = 1.9999999980394758, b = 8.312333324872471e-09


모범 답안)<br>
m = 1.9999999980394758, b = 8.312333324872471e-09


### **2.5) 2차 다항 회귀를 위한 Mini batch 경사하강법 에제 코드입니다. 빈칸을 채워주세요. (7점)**

In [21]:
import numpy as np

X = 2 * np.random.rand(100, 1)
y = 4 + 3 * X + 2 * X**2 + np.random.randn(100, 1)

X_poly = np.c_[X, X**2]

theta = np.random.randn(3, 1)
learning_rate = 0.1
iterations = 1000
batch_size = 20
m = len(X_poly)

def mini_batch_gradient_descent(X_b, y, theta, learning_rate, iterations, batch_size):
    for iteration in range(iterations):
        shuffled_indices = np.random.permutation(m)
        X_b_shuffled = X_b[shuffled_indices]
        y_shuffled = y[shuffled_indices]

        for i in range(0, m, batch_size):
            X_i = X_b_shuffled[i:i+batch_size]
            y_i = y_shuffled[i:i+batch_size]
            gradients = 2/batch_size * X_i.T.dot(X_i.dot(theta) - y_i)
            theta = theta - learning_rate * gradients
    return theta

X_b = np.c_[np.ones((m, 1)), X_poly]

theta_final = mini_batch_gradient_descent(X_b, y, theta, learning_rate, iterations, batch_size)

print("최종 theta 값:", theta_final)

최종 theta 값: [[3.93512952]
 [2.78872744]
 [1.99632765]]


모범 답안)<br>
최종 theta 값: [[4.02673618]<br>
 [3.18859894]<br>
 [1.9938976 ]]